crates/ruff/src/rules/ruff/rules/explicit_f_string_type_conversion.rs

```diff
@@ -39,59 +47,124 @@ impl AlwaysAutofixableViolation for ExplicitFStringTypeConversion {
+    };
+    expression.codegen(&mut state);
+
+    Ok(Fix::automatic(Edit::range_replacement(
+        state.to_string(),
+        range,
+    )))
+}
+
 /// RUF010
 pub(crate) fn explicit_f_string_type_conversion(
     checker: &mut Checker,
-    formatted_value: &Expr,
-    conversion: ConversionFlag,
+    expr: &Expr,
+    values: &[Expr],
 ) {
-    // Skip if there's already a conversion flag.
-    if !conversion.is_none() {
-        return;
-    }
+    let mut formatted_values: Vec<(usize, TextRange)> = vec![];
+    for (index, formatted_value) in values.iter().enumerate() {
+        let Expr::FormattedValue(ast::ExprFormattedValue {
+            conversion,
+            value,
+            ..
+        }) = &formatted_value else {
+            continue;
+        };
+        // Skip if there's already a conversion flag.
+        if !conversion.is_none() {
+            return;
+        }
 
-    let Expr::Call(ast::ExprCall {
-        func,
-        args,
-        keywords,
-        range: _,
-    }) = formatted_value else {
-        return;
-    };
+        let Expr::Call(ast::ExprCall {
+            func,
+            args,
+            keywords,
+            ..
+        }) = value.as_ref() else {
+            return;
+        };
 
-    // Can't be a conversion otherwise.
-    if args.len() != 1 || !keywords.is_empty() {
-        return;
-    }
+        // Can't be a conversion otherwise.
+        if args.len() != 1 || !keywords.is_empty() {
+            return;
+        }
 
-    let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
-        return;
-    };
+        let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
+            return;
+        };
 
-    let conversion = match id.as_str() {
-        "ascii" => 'a',
-        "str" => 's',
-        "repr" => 'r',
-        _ => return,
-    };
+        if !matches!(id.as_str(), "str" | "repr" | "ascii") {
+            return;
+        };
+
+        if !checker.ctx.is_builtin(id) {
+            return;
+        }
+        formatted_values.push((index, value.range()));
+    }
 
-    if !checker.ctx.is_builtin(id) {
+    if formatted_values.is_empty() {
         return;
     }
 
-    let formatted_value_range = formatted_value.range();
-    let mut diagnostic = Diagnostic::new(ExplicitFStringTypeConversion, formatted_value_range);
+    let mut diagnostics = formatted_values
+        .iter()
+        .map(|(_, range)| Diagnostic::new(ExplicitFStringTypeConversion, *range))
+        .collect::<Vec<_>>();
 
-    if checker.patch(diagnostic.kind.rule()) {
-        let arg_range = args[0].range();
-        let remove_call = Edit::deletion(formatted_value_range.start(), arg_range.start());
-        let add_conversion = Edit::replacement(
-            format!("!{conversion}"),
-            arg_range.end(),
-            formatted_value_range.end(),
+    if checker.patch(diagnostics[0].kind.rule()) {
+        let fix = fix_explicit_f_string_type_conversion(
```

Comment:
I think we should generate a separate fix for each diagnostic. As-is, the diagnostic only highlights one violation (correct), but the fix shows modifying the whole string.